### PR TITLE
Convert to Glimmer components, drop block form support

### PR DESCRIPTION
**Pull request by ijlee2**
## Description

This PR addresses and can close https://github.com/shipshapecode/ember-prism/issues/84.

There are a few key changes:

- Bug fix: splattributes are applied to the `<code>` element created by `<CodeBlock>` or `<CodeInline>`
- Feature: `<CodeBlock>` allows the boolean argument `@showLineNumbers` (motivated by https://github.com/ember-learn/ember-octane-vs-classic-cheat-sheet/issues/39)
- Breaking change: `<CodeBlock>` and `<CodeInline>` drop their block form support, in favor of always using the `@code` argument

As a part of the solution, @rwwagner90 and I converted the components to Glimmer components, refactored them, and updated tests. The minimal Ember compatibility version has been increased from 3.12 to 3.13 in order to support component colocation in an addon.

### CODE REVIEW

**Comment by rwwagner90** [comment]
@ijlee2 I added a `code-base` component to extend from, since our components code was identical. I also removed all the old code in favor of the new stuff and fixed the acceptance test. It looks like the only thing left to fix is the test checking if the code updates. It does not seem to update when the content is updated. Mind taking a look?

**Comment by rwwagner90** [comment]
@ijlee2 I ultimately decided to remove support for the block form, and require passing `code` all the time.

**Comment by rwwagner90** [comment]
@ijlee2 any ideas why 3.12 is failing? I'm not sure what version started supporting colocated components.

**Comment by ijlee2** [comment]
Hi, @rwwagner90. Thanks for reaching out to me for feedback and continuing the work on converting to Glimmer components! I saw updates to tests and readme as well, much appreciate your working on them.

I'm not too familiar with addon history (only recently got into making one). From a [recent Discord chat](https://discordapp.com/channels/480462759797063690/608346628163633192/730851857072652298), it looks like colocation works for addons starting 3.13. Can you try updating the `ember-try` scenario from 3.12 to 3.13?

With regards to your two recent commits, I think I'd provide three feedback:

1\. I agree that we could practice inheritance for `<CodeBlock>` and `<CodeInline>`, but only because the backing classes are _currently_ the same. In general, for components, I think I'd prefer to avoid inheritance in favor of duplication or composition. In apps, at least (I lack experience with addons), I feel like a business requirement for a component can change often, so it's better to avoid inheriting early.

2\. I'm personally okay with removing the block form for `<CodeBlock>` and making a breaking change. Off the top of my head, I'm not sure how to listen to a change in the block content and recreate the `<div>` element (`this.blockElement`) in order to recalculate `this.code`. (That's a lot of side effects! 😄) Hopefully, not many people are using the block form. Asking people to migrate to always using `@code` seems reasonable to me.

3\. For `<CodeBlock>`, I think it may be better to apply `...attributes` to the `<code>` element. Currently, I can't think of a use case where it's better to apply style to the `<pre>` element. If we apply `...attributes` to `<code>`, we could reuse `<CodeInline>`:

```handlebars
{{!-- addon/components/code-block.hbs --}}
<pre
  class="code-block {{this.languageClass}} {{if @showLineNumbers "line-numbers"}}"
>
  {{~! ~}}
  <CodeInline
    @code={{@code}}
    @language={{@language}}
    ...attributes
  />
  {{~! ~}}
</pre>
```

The composition would also allow us to remove inheriting `CodeBase` class.

**Comment by rwwagner90** [comment]
@ijlee2 thanks for that detailed feedback! I applied your suggestions about reusing the `<CodeInline/>` component inside the `<CodeBlock/>` template and removing the base class.

Are we okay with dropping support for Ember < 3.13? I suppose it's close enough to the 3.12 default.

**Comment by ijlee2** [comment]
Hmm, it's up to you. I think there's merit to keeping the old component structure, so that you can support 3.12 LTS users for a few more months (based on https://emberjs.com/releases/lts/, it seems 3.12 support is dropped on October 1).

File: addon/components/code-block.js [hunk 1]
```diff
@@ -1,12 +1,11 @@
-import CodeBase from './code-base';
-import layout from '../templates/components/code-block';
+import Component from '@glimmer/component';
 
-export default CodeBase.extend({
-  layout,
-  tagName: 'pre',
-  classNames: ['code-block'],
+export default class NewCodeBlockComponent extends Component {
```
Comment (ijlee2) [inline addon/components/code-block.js:3]:
```suggestion
export default class CodeBlockComponent extends Component {
```

**Comment by rwwagner90** [comment]
@ijlee2 I would be fine moving components to the old structure, but I tried that and it didn't work either. Maybe I did something wrong. Want to take a stab at moving back to classic structure?

**Comment by ijlee2** [comment]
@rwwagner90 Yep, if I have time before Wednesday evening (CST), I'll have a look. If not, please feel free to go with 3.13.

**Comment by ijlee2** [comment]
I tried reverting the component folder structure and export statements in `app/components`, but the tests (for `<CodeBlock>` only, maybe?) continued to fail when I ran `ember try:one ember-lts-3.12`.

I now think it's okay to up the minimum support version to 3.13 while we make a breaking change with dropping block form support.

**Comment by rwwagner90** [comment]
Thanks @ijlee2. I agree. No point in spending a ton of time supporting 3.12.

**Pull request merged**